src/commands/watch.rs

```diff
@@ -185,48 +175,49 @@ impl WatchCommand {
             .await?
             .into_v1();
 
-        let path_patterns: Vec<String> = match self.watch_assets {
-            // Watch patterns
-            false => app_manifest
-                .components
-                .iter()
-                .filter_map(|c| c.build.as_ref())
-                .filter_map(|b| b.watch.clone())
-                .flatten()
+        let watch_patterns = app_manifest
+            .components
+            .iter()
+            .filter_map(|c| c.build.as_ref())
+            .filter_map(|b| b.watch.clone())
+            .flatten();
+        let component_source_patterns = app_manifest
+            .components
+            .iter()
+            .filter_map(|c| {
+                if let RawModuleSource::FileReference(path) = &c.source {
+                    return path.to_str();
+                }
+                None
+            })
+            .map(String::from);
+        let component_file_patterns = app_manifest
+            .components
+            .iter()
+            .filter_map(|c| c.wasm.files.as_ref())
+            .flatten()
+            .map(|raw_file_mount| match raw_file_mount {
+                RawFileMount::Placement(raw_directory_placement) => String::from(
+                    raw_directory_placement
+                        .source
+                        .join("**/*")
+                        .to_str()
+                        .expect("conversion to str not to fail"),
+                ),
+                RawFileMount::Pattern(pattern) => pattern.to_string(),
+            });
+        let path_patterns: Vec<String> = match self.skip_build {
+            false => watch_patterns.chain(component_file_patterns).collect(),
```

Comment:
I apologise for blundering in late and querying the behaviour, but this seems to lump all patterns together, whereas only the `build.watch` patterns need to trigger a rebuild; if a user only changes an asset, we could just restart without rebuilding, right?  But consolidating the list seems to mean we won't be able to tell whether an event falls into the "source code change" or "artifact only change" category...?

(Again, forgive me if this was all dealt with in the earlier PR.)

Reply:
You're correct that it lumps them all together. Dynamically deciding whether `spin build` needs to run or not was taken out of scope in the original PR. In my opinion this is not a P0 feature and is something we can explore as an improvement in the future.